simulator/modules/branch/branch.h

```diff
@@ -42,9 +42,32 @@ class Branch : public Log
     public:
         explicit Branch( bool log);
         void clock( Cycle cycle);
-        auto get_mispredict_rate() const { return ( num_branches != 0) 
-                                                  ? num_mispredictions / num_branches * 100
-                                                  : 0; }
+        auto get_target_mispredictions_num() const { return num_target_mispredictions; }
+        auto get_direction_mispredictions_num() const { return num_direction_mispredictions; }
+        auto get_mispredictions_num() const { return num_direction_mispredictions + num_target_mispredictions; }
+
+        bool is_misprediction( const Instr& instr, const BPInterface& bp_data)
+        {
+            bool is_misprediction = false;
+
+            if ( !bp_data.is_taken)
+            {
+                num_direction_mispredictions++;
+
+                is_misprediction |= instr.is_common_branch() && instr.is_taken();
+
+                is_misprediction |= instr.is_likely_branch() && !instr.is_taken();
```

Comment:
I don't get why you consider that as a mispredict. I don't get why you care about type of branch here as well.
Why can't you just leave the original version of code? It worked well...

Reply:
Let's consider that BPU predicted that jump is not taken, but it was 'likely' branch which we have taken. So if actual state of this branch is 'taken', then it's not misprediction.
But code below counts this situation as mispredict.
`is_misprediction =  instr.get_bp_data().is_taken != instr.is_taken();`

Reply:
> So if actual state of this branch is 'taken', then it's not misprediction.

No, it is. BPU predicted that branch is not taken and PC + 4, PC + 8 ... instructions are fetched. The actual state is taken, and Target, Target + 4 must be fetched.

> But code below counts this situation as mispredict.
> ```is_misprediction = instr.get_bp_data().is_taken != instr.is_taken();```

And that is perfectly correct.

Reply:
Understood. Thanks.